pil-analyzer/src/traits_processor.rs

```diff
@@ -0,0 +1,166 @@
+use core::panic;
+use std::collections::{HashMap, HashSet};
+
+use powdr_ast::{
+    analyzed::{
+        Expression, FunctionValueDefinition, PolynomialReference, Reference, Symbol,
+        TypedExpression,
+    },
+    parsed::{types::Type, FunctionCall, TraitImplementation},
+};
+
+pub struct TraitsProcessor<'a> {
+    definitions: &'a mut HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    implementations: &'a HashMap<String, Vec<TraitImplementation<Expression>>>,
+    type_args_stack: Vec<Vec<Type>>,
+    stack: Vec<String>,
+    entry_point: Option<FunctionValueDefinition>,
+}
+
+impl<'a> TraitsProcessor<'a> {
+    pub fn new(
+        definitions: &'a mut HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+        implementations: &'a HashMap<String, Vec<TraitImplementation<Expression>>>,
+    ) -> Self {
+        Self {
+            definitions,
+            implementations,
+            type_args_stack: Vec::new(),
+            stack: Vec::new(),
+            entry_point: None,
+        }
+    }
+
+    pub fn traits_resolution(&mut self) {
+        let keys = self
+            .definitions
+            .iter()
+            .filter(|(_, (_, def))| {
+                def.is_some()
+                    && !matches!(
+                        def.as_ref().unwrap(),
+                        FunctionValueDefinition::TraitFunction(_, _)
+                    )
+            })
+            .map(|(name, _)| name.clone())
+            .collect::<Vec<_>>();
+        for name in keys {
+            self.dfs_traits(&name);
+
+            self.stack.clear();
+            self.type_args_stack.clear();
+        }
+    }
+
+    fn dfs_traits(&mut self, current: &str) {
+        self.stack.push(current.to_string());
+
+        let next_name = {
+            let current_def = &self.definitions.get(current).unwrap().1;
+            match current_def {
+                Some(FunctionValueDefinition::Expression(TypedExpression {
+                    e: Expression::FunctionCall(_, FunctionCall { function, .. }),
+                    ..
+                })) => match function.as_ref() {
+                    Expression::Reference(
```

Comment:
You can use the following here:
```suggestion
                    e,
                    ..
                })) => e.all_children().filter_map(|e| match e { Expression::Reference(_, Reference::Poly(PolynomialReference{name, type_args: Some(types)})) => Some((name.clone(), types.clone()), _=> None })
```

Reply:
We can also take a look at this together at some point. If this is too tricky, you can also just do fully resolved trait function calls first and we do this one as another PR.

Reply:
> You can use the following here:

But, with this change, wouldn't we be losing the "reference path" to the trait function?

My original idea was to jump from reference to reference waiting for the last reference to jump to a TraitFunction (otherwise rollback and continue searching in other branches) and there have all the necessary type_args.

With this change we would be keeping all the references related to the initial expression, those that go to a TraitFunction and those that don't, in pre-order, which I understand would make more difficult the idea of "finding the reference path".

But maybe I misunderstood the original idea?